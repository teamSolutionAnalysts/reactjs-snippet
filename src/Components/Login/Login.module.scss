.login-block {
  background-color: #ffffff;
  border-radius: 20px;
  -webkit-box-shadow: 0 0 10px rgba(0, 0, 0, 0.55);
  -moz-box-shadow: 0 0 10px rgba(0, 0, 0, 0.55);
  -o-box-shadow: 0 0 10px rgba(0, 0, 0, 0.55);
  -ms-box-shadow: 0 0 10px rgba(0, 0, 0, 0.55);
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.55);
  width: 100%;
  max-width: 570px;
  padding: 50px 60px 10px;
  margin-left: 8%;
}
@media (max-width: 767px) {
  .login-block {
    padding: 30px;
    margin-left: 0;
  }
}
.login-block h2 {
  font-size: 1.875em;
  color: #000000;
  font-weight: 700;
  margin-bottom: 15px;
}
.login-block h3 {
  font-size: 1.125em;
  font-weight: 500;
  margin-bottom: 40px;
}
.forgot-password {
  margin: 8px 8px 0 0;
  display: inline-block;
  vertical-align: top;
  font-size: 0.75em;
  color: #383838;
  font-style: italic;
  cursor: pointer;
}
.forgot-password:hover {
  color: #ffc206;
}
.space-section {
  height: 115px;
}
.or-section {
  text-align: center;
  position: relative;
  margin-top: 15px;
}
.or-section label {
  background-color: #ffffff;
  z-index: 1;
  position: relative;
  padding: 0 15px;
  color: #000000;
  font-weight: 700;
  font-size: 1em;
}
.login-with-otp {
  cursor: pointer;
  margin: 0 !important;
  color: #000000 !important;
  font-size: 0.875em !important;
  font-weight: 700 !important;
  box-shadow: 0px 10px 20px rgb(0 0 0 / 10%) !important;
  border-radius: 30px !important;
  display: block !important;
  padding: 10px;
  border: 1px solid #d3dae3;
  width: 100%;
}
.login-with-otp:hover {
  color: #ffc206;
}
.otp-time {
  margin: 8px 8px 0px 10px;
  font-size: 0.875em;
  color: #000;
}
.otp-time a {
  cursor: pointer;
}
.otp-time a:hover {
  color: #ffc206;
}
.login-btn {
  text-transform: unset;
}
.login-from .btn-large {
  color: #000000;
  margin-top: 15px;
  width: 100%;
  text-transform: capitalize;
}
.login-bottom {
  margin-top: 10px;
  font-size: 1.313em;
}
.login-bottom p {
  font-size: 0.875em;
}
.login-bottom a {
  color: #000;
  font-size: 0.875em;
  font-weight: 500;
  display: inline-block;
  vertical-align: top;
  margin: 0 10px;
}
.login-bottom a:hover {
  color: #ffc206;
}
.login-from .custom-check label::after {
  width: 20px;
  height: 20px;
  top: 4px;
}
.login-from .custom-check label::before {
  width: 13px;
  height: 4px;
  position: absolute;
  left: 4px;
  top: 10px;
  border-left: 1px solid #000;
  border-bottom: 1px solid #000;
}
.login-from .custom-check {
  padding-left: 36px;
}
.login-from .submit-area {
  padding: 0 10px;
}
.custom-check input[type="checkbox"]:checked + label::after {
  background-color: transparent;
  border-color: #000;
}
.form-control,
input[type="text"].form-control {
  display: flex;
  width: 100%;
  padding: 10px 20px;
  font-size: 0.875em;
  color: #000;
  background-color: transparent;
  background-clip: padding-box;
  border: 1px solid #7f7f7f;
  border-radius: 30px;
  height: 45px;
  font-weight: 500;
}
.form-control:focus,
input[type="text"].form-control:focus {
  border: 1px solid #7f7f7f;
  background-color: #fff;
  outline: 0;
  box-shadow: none;
}
.form-control::-webkit-input-placeholder {
  color: #7f7f7f;
  opacity: 1;
}
.form-control::-moz-placeholder {
  color: #7f7f7f;
  opacity: 1;
}
.form-control:-ms-input-placeholder {
  color: #7f7f7f;
  opacity: 1;
}
.form-control::-ms-input-placeholder {
  color: #7f7f7f;
  opacity: 1;
}
.form-control::placeholder {
  color: #7f7f7f;
  opacity: 1;
}
.eyeopen{
  width: 20px;
  position: absolute;
  right: 15px;
  top: 48px;
  cursor: pointer;
  z-index: 9999999;
}
.password-cover{
  position: relative !important;
}

.btn {
  display: inline-block;
  font-weight: bold;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: 1px solid #ffc207;
  background-color: #ffc207;
  color: #000;
  padding: 0 45px;
  font-size: 1em;
  border-radius: 30px;
  cursor: pointer;
  transition: all ease 0.3s;
  text-transform: capitalize;
}
.btn:hover,
.btn:focus {
  text-decoration: none;
  background-color: #000;
  color: #fff !important;
  border-color: #000;
}
.btn:focus,
.btn.focus {
  outline: 0;
  box-shadow: none;
}
.btn-large {
  height: 45px;
  padding: 0 75px;
  font-size: 1em;
  border-radius: 30px;
  width: 100%;
  position: relative;
}
.btn-medium {
  height: 44px;
  line-height: 44px;
  padding: 0 30px;
}
.btn-shadow {
  box-shadow: 0 0 10px rgba(0, 0, 0, 0.35);
  -webkit-box-shadow: 0 0 10px rgba(0, 0, 0, 0.35);
  -ms-box-shadow: 0 0 10px rgba(0, 0, 0, 0.35);
  -moz-box-shadow: 0 0 10px rgba(0, 0, 0, 0.35);
  -o-box-shadow: 0 0 10px rgba(0, 0, 0, 0.35);
}
.btn-yellow-border {
  background-color: transparent;
  border: 1px solid #a9a8a7;
  text-transform: capitalize;
  font-weight: normal;
}
.btn-yellow-border:hover {
  border-color: #000;
}
.otp-section {
  // display: flex;
  align-items: center;
}
.otp-section .form-control1 {
  border-bottom: 1px solid #000000 !important;
  background-color: transparent;
  border-radius: 0;
  border: none;
  text-align: center;
  margin-right: 30px;
  padding: 10px;
  line-height: normal;
  font-size: 1em;
  font-weight: 400;
  width: 20%;
}
.otp-section .form-control1:focus {
  outline: none;
}
.otp-section .form-control1::placeholder {
  font-size: 1.125em;
  font-weight: 400;
}
.otp-section .form-control1:last-child {
  margin-right: 0;
}
.otp-section .form-control1:focus {
  border: transparent !important;
  border-bottom: 1px solid black !important;
}

.otp-section .otpContainer {
  margin: 5% auto;
}

.otp-section .otpInput {
  width: 2.2rem !important;
  height: 2.5rem;
  margin: 0 0.5rem;
  font-size: 1rem;
  text-align: center;
  border-radius: 4px;
  border: none;
  border-bottom: 1px solid #000000;
  border-radius: 0;
  @media (max-width:479px){
    width: 1.6rem !important;
  }
}

.otp-section .otpInput:focus{
  outline: none;
}

.otp-note {
  color: #f52e2e;
  font-size: 1.125em;
  font-style: italic;
  margin-top: 10px;
  margin-bottom: 40px;
}
.resend-otp-section {
  margin-bottom: 50px;
  margin-top: 25px;
}
.resend-otp-section a {
  font-weight: 700;
  color: #000000;
  cursor: pointer;
}
.backto-login-link {
  font-size: 1.25em;
}
.backto-login-link p {
  display: flex;
  align-items: center;
  justify-content: center;
}
.backto-login-link a {
  color: #383838;
  font-weight: 700;
}
.otp-successfull-txt {
  color: #53bb46;
  font-size: 1.125em;
  font-style: italic;
  margin-top: 10px;
  margin-bottom: 40px;
}
.invalid-otp-txt {
  color: #f52e2e;
  font-size: 1.125em;
  font-style: italic;
  margin-top: 10px;
  margin-bottom: 25px;
}

.arabic-body .h2 .h3 {
  text-align: center;
}

.arabic-body .h2 .h3 {
  text-align: center;
}

.devClassName {
  position: absolute;
  height: 100%;
  width: 100%;
  justify-content: center;
  align-items: center;
  display: flex;
}
.devClassName .loaderClassName {
  font-size: 5em;
  color: #ffc206;
}

.smallLoader {
  color: #000;
  position: absolute;
  left: 0px;
  right: 10rem;
  cursor: default;
  top: 7px;
}

.btn:disabled {
  cursor: not-allowed;
  opacity: 0.5;
}

.btn:disabled:hover {
  background: #ffc206;
  color: #000 !important;
}

.disabled{
  opacity: 0.5;
  cursor: not-allowed !important;
  > a {
    pointer-events: none !important;
  }
}

.optContainer{
  width: 100%;
  display: flex;
  justify-content: space-between;
  flex-wrap: wrap;
}

.invalid-username-txt {
  color: #ff4d4f;
  font-size: 1rem;
  font-weight: 400;
  display: flex;
  margin-top: 10px;
}

.valid-username-txt {
  color: #53bb46;
  font-size: 1rem;
  font-weight: 400;
  display: flex;
  margin-top: 10px;
}
