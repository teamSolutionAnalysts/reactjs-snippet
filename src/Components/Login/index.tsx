import React, { useState } from "react";
import { NextPage } from "next";
import { Formik, Form } from "formik";
import Cookie from "js-cookie";
import { useRouter } from "next/router";
import { withTranslation } from "@Server/i18n";
import { STLInput, Loader, STLButton, STLCol, STLRow } from "@Components/Common";
import AuthService from "@Services/API/Auth";
import { LOCAL_HOST_KEY } from "@Utils/enums";
import RoutPath from "@Utils/routes";
import { useDispatch, useSelector } from "react-redux";
import { loginSchema } from "@Utils/validationSchema";
import { getCookie, NotificationWithIcon } from "@Utils/helper";
import ConfigService from "@Services/API/Config";
import CartService from "@Services/API/Cart";
import CartActions from "@Redux/Actions/CartActions";
import AnonymousUserCartActions from "@Redux/Actions/AnonymousUserCartActions";
import PlanActions from "@Redux/Actions/PlanActions";
import loaderStyle from "@Components/Common/Loader.module.scss";
import _ from "lodash";
import Link from "next/link";
import Modal from "react-bootstrap/Modal";
import profileStyle from "@Pages/profile-dashboard/profileDashboard.module.scss";
import style from "./Login.module.scss";

interface Props {
    handleRequestOTPLogin: any;
    getUserDetailsHandler: any;
}

const Login: NextPage<Props> = ({ t, i18n, handleRequestOTPLogin, getUserDetailsHandler }: any) => {
    const router: any = useRouter();
    const dispatch = useDispatch();
    const { AUTH_TOKEN, REFRESH_TOKEN } = LOCAL_HOST_KEY;
    const { Checkout, RootRoute } = RoutPath;
    const [errors, setErrors] = useState<any>(null);
    const [loader, setLoader] = useState<boolean>(false);
    const [isUserNameTouched, setUserNameTouched] = useState(false);
    const commonState = useSelector((state: any) => state.common);
    const { planID } = commonState;
    
    // function to check where to redirect after login page 
    const handleRedirectTo = (): any => {
        if (router?.query?.from === "header" || !router?.query?.from) {
            return RootRoute;
        }
        if (router?.query?.from === "cart") {
            return Checkout;
        }
        if (router?.query?.from === "addon") {
            return `/plan/${planID.planID}`;
        }
        return router?.query?.from ? decodeURIComponent(router?.query?.from) : "/";
    };
    const redirectTo = handleRedirectTo();
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const [userNameRequiredError, setUserNameRequiredError] = useState<any>(null);
    const [linkConfig, setLinkConfig] = React.useState<any>(null);
    const [mobileActiveError, setMobileActiveError] = useState<any>(null);
    const [MSISDNExp, setMSISDNExp] = useState<any>("");
    const [OTPLoginAccess, setOTPLoginAccess] = useState<any>("");
    const [isRequestOTP, setIsRequestOTP] = useState<any>("");
    const [enteredUsername, setUsername] = useState<any>(null);
    const [cartMergeMsg, setCartMergeMsg] = useState<any>(null);
    const [showConfirm, setShowConfirm] = useState<boolean>(false);
    const cartState = useSelector((state: any) => state.cart);
    const anonymousUserState = useSelector((state: any) => state.anonymousUser);
    const { shoppingCartId } = cartState;
    const { shoppingCartId: anonymousshoppingCartId } = anonymousUserState;

    // function to check the credential and set the data which is required after login
    const handleLoginSubmit = async (values: any) => {
        if ((getCookie("uuid") && shoppingCartId) || (!getCookie("uuid") && anonymousshoppingCartId)) {
            // eslint-disable-next-line no-param-reassign
            values = { ...values, cartId: anonymousshoppingCartId || shoppingCartId };
        }
        setLoader(true);
        AuthService.postLoginDetail(values)
            .then(async (res: any) => {
                if (res) {
                    if (res) {
                        Cookie.set(AUTH_TOKEN, res?.token);
                        Cookie.set(REFRESH_TOKEN, res?.refreshToken);
                        const userData = await getUserDetailsHandler();
                        Cookie.remove("uuid");
                        // setting the various data that we get after successful login
                        dispatch(AnonymousUserCartActions.Reset());
                        const createCartResponse: any = await CartService.createTMFShoppingCart({
                            userId: userData?.id,
                            name: userData?.username,
                        });
                        dispatch(
                            CartActions.SetSubscriptionLimit({
                                checkSubscriptionLimitExceed: res?.checkSubscriptionLimitExceed || false,
                            })
                        );
                        dispatch(
                            CartActions.SetAmountOverDueForCustome({
                                isAmountOverDueForCustomer: res?.isAmountOverDueForCustomer || false,
                            })
                        );
                        dispatch(
                            CartActions.SetSpendingLimit({
                                isSpendingLimitExceeded: res?.isSpendingLimitExceeded || false,
                            })
                        );
                        dispatch(
                            CartActions.SetIsRemainingSpendingLimitFound(res?.isRemainingSpendingLimitFound || false)
                        );
                        dispatch(
                            CartActions.SetShoppingCartId({
                                shoppingCartId: createCartResponse?.result?.shoppingCartId,
                            })
                        );
                        if (createCartResponse?.result?.cart?.cartItems?.length > 0) {
                            dispatch(
                                CartActions.setCartBadge({
                                    showBadge: true,
                                    totalItems: createCartResponse?.result?.cart?.cartItems?.length,
                                })
                            );
                            dispatch(
                                PlanActions.SetPaymentMode({
                                    paymentMode: createCartResponse?.result?.cart?.cartItems[0]?.product?.["@type"],
                                })
                            );
                        }
                        if (res?.merged === false && typeof res?.isCartAbleToMerge === "string") {
                            setCartMergeMsg(res?.isCartAbleToMerge);
                            setShowConfirm(true);
                        } else {
                            await router.push(redirectTo);
                        }
                    }
                    setLoader(false);
                }
            })
            .catch(err => {
                setLoader(false);
                setErrors(err?.data?.responseMessage);
            });
    };

    // function to check link of forget password 
    const handleLinkClick = (reqType: any, username: any) => {
        setLoader(true);
        const reqData = {
            requestType: reqType,
            username,
        };
        AuthService.forgotPassword(reqData)
            .then((res: any) => {
                setLoader(false);
                if (res) {
                    NotificationWithIcon("success", res?.responseMessage || "");
                }
            })
            .catch((err: any) => {
                setLoader(false);
                NotificationWithIcon("error", err?.data?.responseMessage || "");
            });
    };

    // function to redirect at forget password page if link is valid or not 
    const handleForgotClick = (username: any) => {
        if (username) {
            if (linkConfig.length === 1 && linkConfig[0] === "Link") {
                handleLinkClick("link", username);
            } else {
                router.push(`/forgotPassword?username=${username}`);
            }
            setUserNameRequiredError(null);
        } else {
            setUserNameRequiredError(t("login:USERNAME_REQUIRED_FOR_FORGOT"));
        }
    };

    // function to request OTP for login
    const handleRequestOTP = (username: any) => {
        if (username) {
            setUserNameRequiredError(null);
            handleRequestOTPLogin(username);
        } else {
            setUserNameRequiredError(t("login:USERNAME_REQUIRED_FOR_REQUEST_OTP"));
        }
    };

    // function to get configuration from API
    const getConfigurations = async () => {
        setLoader(true);
        const configResponse: any = await ConfigService.getConfiguration("");
        setLoader(false);
        if (configResponse?.responseCode === 200) {
            const forgotPasswordOptions = _.find(
                configResponse?.result,
                item => item.configurableKey === "user.forgotPassword.options"
            );
            const msisdn = _.find(configResponse?.result, item => item.configurableKey === "msisdn.regExp");
            const otpLoginFor = _.find(configResponse?.result, item => item.configurableKey === "user.otp.loginFor");
            setLinkConfig(forgotPasswordOptions?.configurableValue);
            setMSISDNExp(msisdn?.configurableValue);
            setOTPLoginAccess(otpLoginFor?.configurableValue);
        }
    };

    // useEffect to call configuration function at every render of the page
    React.useEffect(() => {
        getConfigurations();
    }, []);

    // function to change the username
    const handleUserNameChange = (e: any) => {
        if (!isUserNameTouched) {
            setUserNameTouched(true);
            return;
        }
        setMobileActiveError(null);
        const { value } = e.target;
        setUsername(value);
        setUserNameRequiredError(null);
        const mobilePattern = new RegExp(MSISDNExp);
        const emailPattern = new RegExp(
            /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
        );
        const isValidPhoneNumber = mobilePattern.test(value);
        const isValidEmail = emailPattern.test(value);
        if (isValidPhoneNumber) {
            setLoader(true);
            AuthService.numberStatus({ mobile: value })
                .then((result: any) => {
                    setLoader(false);
                    if (result.status === true) {
                        setMobileActiveError(null);
                        if (OTPLoginAccess === "Email") {
                            setIsRequestOTP(true);
                        } else {
                            setIsRequestOTP(false);
                        }
                    } else {
                        setMobileActiveError(t("formikError:INACTIVE_MOBILE_NUMBER"));
                    }
                })
                .catch(() => {
                    setLoader(false);
                    setMobileActiveError(t("formikError:INACTIVE_MOBILE_NUMBER"));
                });
        } else if (!isValidEmail && value.length > 0) {
            setMobileActiveError(t("formikError:VALID_MOBILE_NUMBER"));
        } else if (isValidEmail) {
            if (OTPLoginAccess === "MSISDN") {
                setIsRequestOTP(true);
            } else {
                setIsRequestOTP(false);
            }
        } else if (OTPLoginAccess === "Both") {
            setIsRequestOTP(true);
        } else {
            setIsRequestOTP(false);
            setMobileActiveError(null);
        }
    };

    // function to confirm submit
    const handleConfirm = () => {
        setShowConfirm(false);
    };
    // function to redirect after confirm submit
    const handleConfirmTransfer = async () => {
        handleConfirm();
        await router.push(redirectTo);
    };
    return (
        <>
            <div>
            {/* loader component */}
                {loader && (
                    <Loader loaderClassName={loaderStyle.loaderClassName} divClassName={loaderStyle.devClassName} />
                )}
                <h2 className={`${style.h2}`}>{t("login:ALREADY_HAVE_AN_ACCOUNT")}</h2>
                <h3 className={`${style.h3}`}>{t("login:LOGIN_TO_RETRIEVE_YOUR_ACCOUNT_SETTINGS")}</h3>
                {/* Formik form to show form at login screen to get the username and password form user */}
                <Formik
                    initialValues={{
                        username: "",
                        password: "",
                    }}
                    validationSchema={() => loginSchema(t)}
                    onSubmit={handleLoginSubmit}
                >
                    {({ values, setFieldValue }) => (
                        <Form>
                            <div className="form-group">
                            {/* customised input component for username */}
                                <STLInput
                                    className={`${style["form-control"]}`}
                                    placeholder={t("johnmartin@gmail.com")}
                                    type="text"
                                    name="username"
                                    label={t("login:ENTER_EMAIL_PHONE_NUMBER")}
                                    value={values.username}
                                    dataTestId="username"
                                    handleInputChange={(e: any) => {
                                        setFieldValue("username", e.target.value);
                                        setErrors(null);
                                        handleUserNameChange(e);
                                    }}
                                />
                                <div className={`${style["invalid-username-txt"]}`} data-testid="username-error">
                                    {!errors?.username && mobileActiveError}
                                </div>
                                <div className={`${style["invalid-username-txt"]}`} data-testid="username-error">
                                    {!errors?.username && userNameRequiredError}
                                </div>
                            </div>
                            <div className={`${style["password-cover"]}`}>
                                <div className="form-group mb-0">
                                {/* customised input component for password */}
                                    <STLInput
                                        className={`${style["form-control"]}`}
                                        placeholder={t("***************************")}
                                        type="password"
                                        name="password"
                                        label={t("login:PASSWORD")}
                                        value={values.username}
                                        dataTestId="password"
                                        handleInputChange={(e: any) => {
                                            setFieldValue("password", e.target.value);
                                            setErrors(null);
                                        }}
                                    />
                                </div>
                            </div>
                            <div className="form-group">
                                <p className="text-right">
                                    <label
                                        className={`${style["forgot-password"]}`}
                                        onClick={() => handleForgotClick(values.username)}
                                    >
                                        {t("login:FORGOT_PASSWORD")}
                                    </label>
                                </p>
                            </div>
                            <div className="text-center">
                                <button
                                    type="submit"
                                    className={`${style.btn} ${style["btn-large"]}`}
                                    disabled={loader === true}
                                    data-testid="submit-login"
                                >
                                    {loader && <Loader divClassName={style.smallLoader} />}
                                    {t("login:LOGIN")}
                                </button>
                            </div>
                        </Form>
                    )}
                </Formik>
                <div className={`${style["invalid-otp-txt"]}`} data-testid="login-error">
                    {errors}
                </div>

                <div className={`${style["or-section"]}`}>
                    <label>{t("login:OR")}</label>
                </div>
                <div className={`${style["login-bottom"]} text-center`}>
                    <p className="text-center">
                        <button
                            className={`${style["login-with-otp"]}`}
                            disabled={isRequestOTP}
                            onClick={() => handleRequestOTP(enteredUsername)}
                        >
                            {t("login:REQUEST_OTP")}
                        </button>
                    </p>
                    <p>
                        {t("login:DO_NOT_HAVE_AN_ACCOUNT")}
                        <Link
                            href={{
                                pathname: RoutPath.Register,
                                query: { from: router?.query?.from },
                            }}
                        >
                            <a onClick={() => setLoader(true)}>{t("login:REGISTER_HERE")}</a>
                        </Link>
                    </p>
                </div>
            </div>
            {/* Modal start */}
            <Modal show={showConfirm} onHide={handleConfirm} className="changePinModal" centered backdrop="static">
                <Modal.Header className="justify-content-center">
                    <Modal.Title>
                        <img src="/images/alert.svg" alt="" className="w-auto mr-0" />
                    </Modal.Title>
                </Modal.Header>
                <Modal.Body className={profileStyle.modalBody}>
                    <div className={i18n.language === "ar" ? `${profileStyle["arabic-body"]}` : ""}>
                        <>
                            <STLRow gutter={16}>
                                <STLCol xl={24} lg={24} md={24} sm={24} xs={24} className={profileStyle["mb-20"]}>
                                    <div className={`${profileStyle["form-section"]} border-0`}>
                                        <div className={profileStyle["drop-section"]}>
                                            <div className={profileStyle["form-group"]}>
                                                <label>{t(cartMergeMsg)}</label>
                                            </div>
                                        </div>
                                    </div>
                                </STLCol>
                                <STLCol xl={24} lg={24} md={24} sm={24} xs={24} className="d-flex">
                                    <STLButton className={`${profileStyle.yellowBtn} `} onClick={handleConfirmTransfer}>
                                        {t("login:OK")}
                                    </STLButton>
                                </STLCol>
                            </STLRow>
                        </>
                    </div>
                </Modal.Body>
            </Modal>
            {/* Modal end */}
        </>
    );
};

export default withTranslation(["login", "formikError"])(Login);
