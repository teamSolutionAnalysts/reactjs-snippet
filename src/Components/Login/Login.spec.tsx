import React from "react";
import { render, fireEvent, act } from "@Test/utils";

import "@testing-library/jest-dom/extend-expect";

import { Provider } from "react-redux";
import { store } from "@Redux";
import { rest } from "msw";
import { setupServer } from "msw/node";
import { renderWithRouter } from "@Test/mockTest";

import Login from ".";

const fakeUserResponse = { auth_token: "fake_user_token" };

const server = setupServer(
    rest.post(
        "/api/auth/signin",
        (req: any, res: (arg0: any) => any, ctx: { json: (arg0: { auth_token: string }) => any }) => {
            return res(ctx.json(fakeUserResponse));
        }
    )
);

beforeAll(() =>
    server.listen({
        onUnhandledRequest: "warn",
    })
);

afterEach(() => {
    server.resetHandlers();
});

afterAll(() => server.close());

describe("Login render Page", () => {
    it("should match snapshot", () => {
        const { container } = render(
            <Provider store={store}>
                <Login />
            </Provider>
        );
        expect(container).toMatchSnapshot();
    });

    it("render 2 input components", () => {
        const { getByTestId } = render(
            <Provider store={store}>
                <Login />
            </Provider>
        );
        expect(getByTestId("username")).toBeInTheDocument();
        expect(getByTestId("password")).toBeInTheDocument();
    });

    it("renders a submit button", () => {
        const { getByTestId } = render(
            <Provider store={store}>
                <Login />
            </Provider>
        );
        expect(getByTestId("submit-login")).toBeInTheDocument();
    });

    it("allows the user to login successfully", async () => {
        const { getByTestId } = renderWithRouter(
            <Provider store={store}>
                <Login />
            </Provider>
        );

        // fill out the form
        fireEvent.change(getByTestId("username"), {
            target: { value: "hiro@mailinator.com" },
        });
        fireEvent.change(getByTestId("password"), {
            target: { value: "Test@361_879" },
        });

        await act(async () => {
            await fireEvent.click(getByTestId("submit-login"));
        });
    });

    it("handles server exceptions", async () => {
        // mock the server error response for this test suite only.
        server.use(
            rest.post("/api/auth/signin", (req, res, ctx) => {
                return res(ctx.status(401), ctx.json({ code: 401, error: "Invalid user credentials" }));
            })
        );

        const { getByTestId } = renderWithRouter(
            <Provider store={store}>
                <Login />
            </Provider>
        );

        fireEvent.change(getByTestId("username"), {
            target: { value: "hiro@mailinator.com" },
        });
        fireEvent.change(getByTestId("password"), {
            target: { value: "Test@361_879" },
        });

        await act(async () => {
            await fireEvent.click(getByTestId("submit-login"));
        });

        await act(async () => {
            await expect(getByTestId("login-error"));
        });
    });
});
