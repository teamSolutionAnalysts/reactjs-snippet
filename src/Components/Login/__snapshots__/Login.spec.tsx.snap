// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login render Page should match snapshot 1`] = `
<div>
  <div>
    <h2
      class="undefined"
    >
      login:ALREADY_HAVE_AN_ACCOUNT
    </h2>
    <h3
      class="undefined"
    >
      login:LOGIN_TO_RETRIEVE_YOUR_ACCOUNT_SETTINGS
    </h3>
    <form
      action="#"
    >
      <div
        class="form-group"
      >
        <div>
          <div>
            <label>
              login:ENTER_EMAIL_PHONE_NUMBER
            </label>
          </div>
          <input
            class="ant-input undefined undefined undefined"
            data-testid="username"
            name="username"
            placeholder="johnmartin@gmail.com"
            type="text"
            value=""
          />
          <span
            class="ant-typography ant-typography-danger undefined undefined"
            data-testid="form-error"
          />
        </div>
        <div
          class="undefined"
        />
      </div>
      <div
        class="undefined"
      >
        <div
          class="form-group mb-0"
        >
          <div>
            <div>
              <label>
                login:PASSWORD
              </label>
            </div>
            <input
              class="ant-input undefined undefined undefined"
              data-testid="password"
              name="password"
              placeholder="***************************"
              type="password"
              value=""
            />
            <span
              class="ant-typography ant-typography-danger undefined undefined"
              data-testid="form-error"
            />
          </div>
        </div>
      </div>
      <div
        class="form-group"
      >
        <p
          class="text-right"
        >
          <label
            class="undefined"
          >
            login:FORGOT_PASSWORD
          </label>
        </p>
      </div>
      <div
        class="text-center"
      >
        <button
          class="undefined undefined"
          data-testid="submit-login"
          type="submit"
        >
          login:LOGIN
        </button>
      </div>
    </form>
    <div
      class="undefined"
      data-testid="login-error"
    />
  </div>
</div>
`;
